### ISSUE TITLE
Provide `readyPromiseReject` to `instantiateWasm`

### ISSUE THREAD

**Issue by FelixNumworks** [emscripten-core/emscripten#23038]
When providing an `instantiateWasm` method to the Module, only a `successCallback` is passed to the method (see [docs here](https://github.com/emscripten-core/emscripten/blob/06cebfca51f908d3bc6897113330e45a77010db2/site/source/docs/api_reference/module.rst?plain=1#L169))

The problem is that no `failureCallback` is provided.
In the case of a sync instantiation of the WASM, I can return `false` if the instantiation fails.
But in the case of an async instantiation, if the instantiation fails, the error isn't properly handled and the page freezes.

I'd like `instantiateWasm` to take 3 args instead of 2: `imports`, `successCallback` and `failureCallback`.

This means changing this line:

https://github.com/emscripten-core/emscripten/blob/06cebfca51f908d3bc6897113330e45a77010db2/src/preamble.js#L1067

to

```js
return Module['instantiateWasm'](info, receiveInstance, readyPromiseReject);
```

### Current workaround

For others having the same problem, I'm currently using a dirty hack through `pre-js.js` by doing:

```js 
// pre-js.js

if (Module['readyPromiseRejectWrapper']) {
  Module['readyPromiseRejectWrapper']['value'] = readyPromiseReject;
}
```

And when I instantiate my module:

```js
const readyPromiseRejectWrapper = { value: (e) => {} };
// Pass the wrapper to Module
const module = Module(
  readyPromiseRejectWrapper,
  // You can use the wrapper in 'instantiateWasm' when instantiation fails
  instantiateWasm: (imports, successCallback) => {
    (async () => {
      try {
        // instantiate the wasm
        successCallback(instance, module);
      } catch (e) {
        readyPromiseRejectWrapper.value(e);
      }
    })();
    return {};
  }
);
```
**Version of emscripten/emsdk:**
emcc 3.1.61

**Comment by edmond-chow** [comment]
Do you mean when I call the .catch or .then using Module({}) after still cannot catch up the things from quit_? Also where can I find the pre-js.js using a dirty hack?

**Comment by FelixNumworks** [comment]
> Do you mean when I call the .catch or .then using Module({}) after still cannot catch up the things from quit_?

Sorry I don't understand the sentence :/ What I mean is that in `instantiateWasm`, there is no way to properly reject the promise after an instanciation failure.

> Also where can I find the pre-js.js using a dirty hack?

You should create the `pre-j.js` file yourself, and add it as an `emcc` argument.
See https://emscripten.org/docs/tools_reference/emcc.html (search `-pre-js` in the page)

**Comment by hoodmane** [comment]
I've definitely run into this problem that if an error is thrown in `instantiateWasm` Emscripten does not handle it cleanly so +1 for this issue.

I think we ought to deprecate `instantiateWasm` and replace it with a new setting that we can pass an async function to. Emscripten can use normal async/await now and that will make both user code and the Emscripten source simpler.

**Comment by sbc100** [comment]
@FelixNumworks, can I ask what you use case is for overriding `instantiateWasm`?   I'm trying to understand how/why certain options are used.

**Comment by FelixNumworks** [comment]
@sbc100

I need to `fetch` my `.wasm`. The canonical method to do this would be to pass a `locateFile` method to the module. The problem is that I need to `fetch` it with the option `{ credentials: 'include' }`, which is not done by default (default is `{ credentials: 'same-origin' }`

I am now realizing that the parameter `fetchSettings` can be passed to the Module (found [here](https://github.com/emscripten-core/emscripten/blob/8686d5b7c7fa64b17ad50040a3341014fc83165a/site/source/docs/api_reference/module.rst#other-methods)), but it needs to be explicitely declared in [`INCOMING_MODULE_JS_API`](https://emscripten.org/docs/api_reference/module.html#compilation-settings)

Since this works (`locateFile` + `fetchSettings`) and results in a clearer implementation, I will probably use this method rather than `instantiateWasm`

I must say though that having to explicitly set all my module parameters in `INCOMING_MODULE_JS_API` is a bit burdensome.

**Comment by sbc100** [comment]
> I must say though that having to explicitly set all my module parameters in `INCOMING_MODULE_JS_API` is a bit burdensome.

Yes, unfortunately is code size and complexity thing.   The problem is that each element of `INCOMING_MODULE_JS_API` adds code size and complexity to the generated program.    We would like to keep the default output simple and small and have folks opt into extra complexity.

**Comment by sbc100** [comment]
Great that `fetchSettings` works for you.    We will still obviously work to fix/improve `instantiateWasm` in case there are other use cases out there.
